includes/classes/prod_cat_pulldown.php

```diff
@@ -0,0 +1,186 @@
+                FROM " . TABLE_PRODUCTS . " p
+                LEFT JOIN " . TABLE_PRODUCTS_DESCRIPTION . " pd USING (products_id)
+                LEFT JOIN " . TABLE_PRODUCTS_TO_CATEGORIES . " ptc USING (products_id) "
+				. $this->attributes_join .
+				" WHERE pd.language_id = " . (int)$_SESSION['languages_id'];
+		}
+
+		private function run_sql()
+		{
+			global $db, $currencies;
+
+			$search = '';
+			if (!empty($this->keywords)) {
+				$search = zen_build_keyword_where_clause($this->keyword_search_fields,
+					trim(zen_db_input(zen_db_prepare_input($this->keywords))));
+			}
+			$this->sql .= $search;
+
+			switch (true) {
+				case ($this->option_filter === -1): // no selection made: do not list any products
+				case ($this->option_filter === ''): // no selection made: do not list any categories
+					// no selection made yet
+					break;
+				case ((int)$this->option_filter > 0) : // an Option Name was selected: show only products using attributes with this Option Name
+					$this->sql .= $this->condition;
+					$this->sql .= $this->sort;
+					break;
+				default: //legacy: show all products with attributes
+					$this->sql .= $this->sort;
+					break;
+			}
+
+			$results = $db->Execute($this->sql);
+
+			foreach ($results as $result) {
+				if (!in_array($result['common_id'], $this->exclude)) {
+					switch (true) {
+						case ($this->pulldown == 'category'):
+							$this->values[] = [
+								'id' => $result['categories_id'],
+								'text' => $this->category_text($result),
+							];
+							break;
+						default:
+							$display_price = zen_get_products_base_price($result['products_id']);
+							$name = zen_get_products_name($result['products_id']);
+							$this->values[] = [
+								'id' => $result['products_id'],
+								'text' => sprintf($this->output_string, trim(zen_clean_html($name)),
+										($this->show_model ? ' [' . $result['products_model'] . '] ' : ''),
+										$currencies->format($display_price)) . ($this->show_id ? ' - ID# ' . $result['products_id'] : ''),
+							];
+							break;
+					}
+				}
+			}
+		}
+
+		private function category_text($category)
+		{
+			if ($this->attributes) {
+				$text = $category['categories_name'];
+				if ($this->show_full_path) {
+					$text = zen_output_generated_category_path($category['categories_id']);
+				}
+			} else {
+				$parent = '';
+				if ($this->show_parent) {
+					$parent = zen_get_categories_parent_name($category['categories_id']);
+					if ($parent != '') {
+						$parent = ' : in ' . $parent;
+					}
+				}
+				$text = $category['categories_name'] . $parent . ($this->show_id ? ' - ID# ' . $category['categories_id'] : '');
+			}
+			return $text;
+		}
+
+		function pull_down(string $name, $parameters = '')
+		{
+			if (empty($parameters)) {
+				$parameters = $this->parameters;
+			}
+			return zen_draw_pull_down_menu($name, $this->values, $this->set_selected, $parameters);
```

Comment:
What about the `$required` parameter?

Reply:
this pr is an attempt to refactor 4 functions: `zen_draw_products_pull_down`, `zen_draw_products_pull_down_attributes`, `zen_draw_products_pull_down_categories`, and `zen_draw_products_pull_down_categories_attributes`.  none of these 4 functions implement that attribute.  so i choose to go with the default of `false`.  i am not opposed to a discussion as to whether these functions need that to be true; but i would have to examine closer.